### PR TITLE
Scale and crop images

### LINKED ISSUE #497: Scale and crop images
We're loading full-size images for the cards, which causes the page to load more slowly than it needs to. 

screenshot from 2018-11-13 09-54-56

Additionally, the max-height fix doesn't always give us the right aspect ratio. We should crop the images instead.
screenshot from 2018-11-13 09-56-26

Let's use Hugo pipelines to automate these adjustments. This is a follow-up to #321 and #308.

### PR DESCRIPTION
**Pull request by rdebeasi**
**What issue does this PR solve?**
We currently load-full-size images for homepage cards. As a result, the homepage is [slow to load](https://www.webpagetest.org/result/181127_1C_32322e249b5f4c43af9d7a6b711f2182/).

Page weight: 33 MB. 

**Explain the problem and the proposed solution**
Hugo pipelines [aren't yet supported in Netlify](https://github.com/netlify/build-image/issues/183
), so we need to scale and crop images ourselves. I've added a step to the build that uses ImageMagick to scale and crop images for use in cards. Resolves #497. 

I've also updated to the latest version of Hugo (0.51) so that we can re-define variables without needing to use `Scratch` (new in 0.48).

Page weight after these changes: about 3 MB.

**Next Steps**
Running the site locally is now a 9-step process, and it probably doesn't work properly on Windows. I'd like to provide a Docker image based on https://github.com/netlify/build-image that can be used to run the site locally with minimal setup.

Once pipes are supported in Netlify, we could run search index generation and image cropping directly in Hugo, which would allow us to go back to using just `hugo` for local development. Using pipes would probably also speed up builds; they take ~29 sec currently and will take ~49 sec once we add image processing. I'd like to keep build times under 1 minute.

I'd also like to move user uploads to their own directory so that `static/images` isn't a mix of subdirectories and image files.

---

@springdo , does this seem like a reasonable approach?

### CODE REVIEW

**Comment by rdebeasi** [comment]
Hm, the deploy preview doesn't seem to have built the images. From [the log](https://app.netlify.com/sites/openpracticelibrary/deploys/5bfdbafa40e2a44a907cb738):

```
4:45:52 PM: ##################################################
4:45:52 PM:     START BUILD STEP:  ./scale.sh
4:45:52 PM: ##################################################
4:45:52 PM: Scaling and cropping *.{jpg,jpeg,png}
4:46:05 PM: \n##################################################\n
4:46:05 PM:     BUILD STEP SUCCESS: ./scale.sh
4:46:05 PM: \n##################################################\n
```

I'd expect to see a list of processed files there, or at least an error message. I'm investigating. Yet another reason for us to use Docker, I guess! ;)

File: build.sh [hunk 1]
```diff
@@ -21,25 +21,28 @@ _printBuildStep(){
...
 }
 
 clean(){
-    _printBuildStep "rm -rvf public"
+    _printBuildStep "rm -rvf public static/images/scaled"
 }
 
 test(){
-    _printBuildStep "echo test-command"
+    _printBuildStep "echo No tests defined"
```
Comment (springdo) [inline build.sh]: Lolz - TODO add tests ;)

**Comment by rdebeasi** [comment]
Thanks for taking a look! I totally agree that it would be better to not have an additional script in the project root. I looked into moving this logic into a function in build.sh, but I ran into a couple of problems:

- `_printBuildStep` doesn't seem to accept shell functions, only commands. We'd need to modify it to take either one.
- All the other functions in `build.sh` are build stages, but this scale function would be a build step within the build stage. It seems like this might get a little confusing.

I could alternately make scaling a build stage, but then we'd wind up with a `_printBuildStep` call for each image, which seems... verbose. Or I could just bypass `_printBuildStep` and output the headers myself, but that's a lot of `echo` statements in one file. :upside_down_face: 

Thoughts? Is there a simple solution that I'm missing here?

**Comment by springdo** [comment]
yeah - non trivial problem that in my eyes. 
I guess i see two solutions, 
1) We ignore the print statement for calling it and you add the image stuff into the `build() { }`
2) You create a function under the `_printBuildStep()` for your `image-combobulation` , then in your build function add a call like `_printBuildStep "./build.sh image-combobulation"`. Probs a bit hacky though right?

**Comment by rdebeasi** [comment]
I actually really like option 2 - to me, bash scripts are _all about_ clever hackery! This approach works on my local machine (Fedora 28), but it seems to fail on the build server:

```
4:45:30 PM: ##################################################
4:45:30 PM:     START BUILD STEP:  ./build.sh scaleImages
4:45:30 PM: ##################################################
4:45:30 PM: ./build.sh: 44: ./build.sh: Bad substitution
4:45:30 PM: ##################################################
4:45:30 PM:     BUILD STEP FAILURE: ./build.sh scaleImages
4:45:30 PM: ##################################################
```

Netlify is running an old version of Ubuntu LTS on their build servers, so I'm inclined to blame that. Anyway, I ended up with option 1, and it works great. Thanks!

**Comment by springdo** [comment]
Grand! Be curious to know why option 1 did that on ubuntu, but it's probs not worth the thinking power.

Also - plus, like a MILLION points for your function description 
🦄 🤤  `Combobulate Images` 💯 🥇

**Comment by jhabdas** [comment]
https://github.com/aFarkas/lazysizes

**Pull request merged**